File: app/early_voting.py
"""EPIC-06 期日前投票を確認できる。"""
import sqlite3

from app.elections import get_election
from app.region_elections import list_elections_for_region


def add_early_voting_place(
    conn: sqlite3.Connection, *, election_id: int, name: str, address: str,
    period_start: str, period_end: str, open_time: str = "08:30", close_time: str = "20:00",
) -> dict:
    get_election(conn, election_id)  # 存在確認
    cur = conn.execute(
        "INSERT INTO early_voting_places (election_id, name, address, period_start, "
        "period_end, open_time, close_time) VALUES (?, ?, ?, ?, ?, ?, ?)",
        (election_id, name, address, period_start, period_end, open_time, close_time),
    )
    conn.commit()
    row = conn.execute(
        "SELECT * FROM early_voting_places WHERE id = ?", (cur.lastrowid,)
    ).fetchone()
    return dict(row)


def list_early_voting_places_for_region(
    conn: sqlite3.Connection, region_id: int, election_id: int
) -> list[dict]:
    """登録地域に関係する選挙について、期日前投票の期間・投票所・住所・受付時間を確認する。"""
    relevant_ids = {e["id"] for e in list_elections_for_region(conn, region_id)}
    if election_id not in relevant_ids:
        raise ValueError(f"この選挙は指定地域に関係しません: election_id={election_id}")
    rows = conn.execute(
        "SELECT * FROM early_voting_places WHERE election_id = ? ORDER BY period_start",
        (election_id,),
    ).fetchall()
    return [dict(r) for r in rows]

File: src/logic/earlyVoting.js
// EPIC-06 期日前投票を確認できる。
import { loadState, saveState, nextId } from "./db.js"
import { getElection } from "./elections.js"
import { listElectionsForRegion } from "./regionElections.js"

export function addEarlyVotingPlace({
  electionId,
  name,
  address,
  periodStart,
  periodEnd,
  openTime = "08:30",
  closeTime = "20:00",
}) {
  const state = loadState()
  getElection(electionId, state) // 存在確認
  const place = {
    id: nextId(state, "early_voting_places"),
    election_id: electionId,
    name,
    address,
    period_start: periodStart,
    period_end: periodEnd,
    open_time: openTime,
    close_time: closeTime,
  }
  state.early_voting_places.push(place)
  saveState(state)
  return place
}

export function getEarlyVotingPlace(placeId, state = loadState()) {
  const place = state.early_voting_places.find((p) => p.id === placeId)
  if (!place) {
    throw new Error(`期日前投票所が見つかりません: id=${placeId}`)
  }
  return place
}

// 管理画面用に、選挙に紐づく期日前投票所を全件確認する（地域による絞り込みなし）。
export function listEarlyVotingPlaces() {
  return [...loadState().early_voting_places].sort((a, b) => (a.period_start < b.period_start ? -1 : 1))
}

export function updateEarlyVotingPlace(placeId, { electionId, name, address, periodStart, periodEnd, openTime, closeTime }) {
  const state = loadState()
  const place = getEarlyVotingPlace(placeId, state)
  getElection(electionId, state) // 存在確認
  place.election_id = electionId
  place.name = name
  place.address = address
  place.period_start = periodStart
  place.period_end = periodEnd
  place.open_time = openTime
  place.close_time = closeTime
  saveState(state)
  return place
}

export function deleteEarlyVotingPlace(placeId) {
  const state = loadState()
  getEarlyVotingPlace(placeId, state) // 存在確認
  state.early_voting_places = state.early_voting_places.filter((p) => p.id !== placeId)
  saveState(state)
}

// 登録地域に関係する選挙について、期日前投票の期間・投票所・住所・受付時間を確認する。
export function listEarlyVotingPlacesForRegion(regionId, electionId) {
  const relevantIds = new Set(listElectionsForRegion(regionId).map((e) => e.id))
  if (!relevantIds.has(electionId)) {
    throw new Error(`この選挙は指定地域に関係しません: election_id=${electionId}`)
  }
  return loadState()
    .early_voting_places.filter((p) => p.election_id === electionId)
    .sort((a, b) => (a.period_start < b.period_start ? -1 : 1))
}
